### PR TITLE
Api keys

### LINKED ISSUE #1867: Adds management of API Keys and their permissions
# Summary

The goal of this issue is to improve the way of handling API keys by following the related specification: https://github.com/meilisearch/specifications/pull/85
(Still a WIP)

# Implementation
## [Actions](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#actions-list-definition)
- [x] `search`  Provides access to GET and POST methods on /indexes/:authorizedIndexes/search routes.
- [x] `documents.add`   Provides access to POST and PUT on /indexes/:authorizedIndexes/documents routes.
- [x] `documents.get`  Provides access to GET methods on /indexes/:authorizedIndexes/documents/ and /indexes/:authorizedIndexes/documents/:documentId routes.
- [x] `documents.delete`    Provides access to DELETE method on indexes/:authorizedIndexes/documents/:documentId, indexes/:authorizedIndexes/documents/:documentId and POST method on indexes/:authorizedIndexes/documents/delete-batch routes.
- [x] `indexes.add`     Provides access to POST /indexes. ⚠️ The newly created index is added to the indexes array for this key and all others keys having [*] as a value for indexes.
- [x] `indexes.get`     Provides access to GET /indexes and /indexes/:authorizedIndexes. ⚠️Non-authorized indexes are ommited from the response on /indexes.
- [x] `indexes.update`  Provides access to PUT /indexes/:authorizedIndexes
- [x] `indexes.delete`  Provides access to DELETE /indexes/:authorizedIndexes
- [x] `tasks.get`   Provides access to GET /tasks and /indexes/:indexUid/tasks routes.
- [x] `settings.get`    Provides access to GET /indexes/:authorizedIndexes/settings and /indexes/:authorizedIndexes/settings/* routes.
- [x] `settings.update`     Provides access to POST /indexes/:authorizedIndexes/settings and /indexes/:authorizedIndexes/settings/* routes.
- [x] `settings.reset`  Provides access to DELETE /indexes/:authorizedIndexes/settings and /indexes/:authorizedIndexes/settings/* routes.
- [x] `stats`   Provides access to GET /stats/. ⚠️Non-authorized indexes are ommited from the response on /stats. Also add access to GET indexes/:authorizedIndexes/stats.
- [x] ` dumps`   Provides access to POST /dumps and GET /dumps/:dumpUid routes.

### [**POST** - `/keys`](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#request-definition)
- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.
- [x] Omitted Content-Type header returns a [missing_content_type](0061-error-format-and-definitions.md#missing_content_type) error.
- [x] Sending an empty Content-Type returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending a different Content-Type than `application/json` returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending an empty payload returns a [missing_payload](0061-error-format-and-definitions.md#missing_payload) error.
- [x] Sending a different payload type than the Content-Type header returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Sending an invalid json format returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Omitting `actions` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Omitting `indexes` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Omitting `expiresAt` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Sending an invalid value for the `actions` field returns an [invalid_api_key_actions](0061-error-format-and-definitions.md#invalid_api_key_actions) error.
- [x] Sending an invalid value for the `indexes` field returns an [invalid_api_key_indexes](0061-error-format-and-definitions.md#invalid_api_key_indexes) error.
- [x] Sending an invalid value for the `expiresAt` field returns an [invalid_api_key_expires_at](0061-error-format-and-definitions.md#invalid_api_key_expires_at) error.
- [x] Sending an invalid value for the `description` field returns an [invalid_api_key_description](0061-error-format-and-definitions.md#invalid_api_key_description) error.

### [**GET** - `/keys/:key`](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#as-anna--i-want-to-get-details-about-an-api-key)

- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.
- [x] Attempting to access an API key that does not exist returns an [api_key_not_found](0061-error-format-and-definitions.md#api_key_not_found) error.

### [**PUT** - `/keys/:key`](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#as-anna--i-want-to-update-api-key-to-change-its-restrictions)

- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.
- [x] Attempting to access an API key that does not exist returns a [api_key_not_found](0061-error-format-and-definitions.md#api_key_not_found) error.
- [x] Omitted Content-Type header returns a [missing_content_type](0061-error-format-and-definitions.md#missing_content_type) error.
- [x] Sending an empty Content-Type returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending a different Content-Type than `application/json` returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending an empty payload returns a [missing_payload](0061-error-format-and-definitions.md#missing_payload) error.
- [x] Sending a different payload type than the Content-Type header returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Sending an invalid json format returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Omitting `actions` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Omitting `indexes` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Omitting `expiresAt` field from the payload returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Omitting `description` field while his value is different than `null` on the server-side returns a [missing_parameter](0061-error-format-and-definitions.md#missing_parameter) error.
- [x] Sending an invalid value for the `actions` field returns an [invalid_api_key_actions](0061-error-format-and-definitions.md#invalid_api_key_actions) error.
- [x] Sending an invalid value for the `indexes` field returns an [invalid_api_key_indexes](0061-error-format-and-definitions.md#invalid_api_key_indexes) error.
- [x] Sending an invalid value for the `expiresAt` field returns an [invalid_api_key_expires_at](0061-error-format-and-definitions.md#invalid_api_key_expires_at) error.
- [x] Sending an invalid value for the `description` field returns an [invalid_api_key_description](0061-error-format-and-definitions.md#invalid_api_key_description) error.

### [**GET** - `/keys`](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#as-anna--i-want-to-list-the-api-keys)

- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](https://github.com/meilisearch/specifications/blob/develop/text/0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](https://github.com/meilisearch/specifications/blob/develop/text/0061-error-format-and-definitions.md#invalid_api_key) error.

### [**DELETE** - `/keys/:key`](https://github.com/meilisearch/specifications/blob/dbefc66902b4c71388020c1e3a2e5cd023c3a0d3/text/0085-api-keys-scoped-api-keys.md#as-anna--i-want-to-delete-an-api-key)

- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](https://github.com/meilisearch/specifications/blob/develop/text/0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](https://github.com/meilisearch/specifications/blob/develop/text/0061-error-format-and-definitions.md#invalid_api_key) error.
- [x] Attempting to access an API key that does not exist returns a `api_key_not_found`.

### [**PATCH** - `/keys/:key`](https://github.com/meilisearch/specifications/blob/6e1af98f817e37da2515803da53599c1b5b74a67/text/0085-api-keys.md#as-anna--i-want-to-update-api-key-to-change-its-restrictions)

- [x] test
- [x] Accessing this route without the `X-MEILI-API-KEY` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- [x] Accessing this route with a key that does not have permissions (i.e. other than the master-key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.
- [x] Attempting to access an API key that does not exist returns a [api_key_not_found](0061-error-format-and-definitions.md#api_key_not_found) error.
- [x] Omitting Content-Type header returns a [missing_content_type](0061-error-format-and-definitions.md#missing_content_type) error.
- [x] Sending an empty Content-Type returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending a different Content-Type than `application/json` returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- [x] Sending an empty payload returns a [missing_payload](0061-error-format-and-definitions.md#missing_payload) error.
- [x] Sending a different payload type than the Content-Type header returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Sending an invalid json format returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- [x] Sending an invalid value for the `actions` field returns an [invalid_api_key_actions](0061-error-format-and-definitions.md#invalid_api_key_actions) error.
- [x] Sending an invalid value for the `indexes` field returns an [invalid_api_key_indexes](0061-error-format-and-definitions.md#invalid_api_key_indexes) error.
- [x] Sending an invalid value for the `expiresAt` field returns an [invalid_api_key_expires_at](0061-error-format-and-definitions.md#invalid_api_key_expires_at) error.
- [x] Sending an invalid value for the `description` field returns an [invalid_api_key_description](0061-error-format-and-definitions.md#invalid_api_key_description) error.

## [Using `API Key` on client-code](https://github.com/meilisearch/specifications/blob/6e1af98f817e37da2515803da53599c1b5b74a67/text/0085-api-keys.md#using-api-key-on-client-code)

- [x] Accessing any route with an `API Key` that has expired or, been deleted returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key)
    - [x] /indexes/../search
    - [x] /indexes/../documents
    - [x] /indexes/../documents/..
    - [x] /indexes/../settings
    - [x] /indexes/../updates
    - [x] /indexes/../stats
    - [x] /indexes/../
    - [x] /indexes
    - [x] /keys/..
    - [x] /keys
- [x] Accessing any route with an API Key that don't have sufficient permissions to access it returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key)
on routes:
    - [x] /indexes/../search
    - [x] /indexes/../documents
    - [x] /indexes/../documents/..
    - [x] /indexes/../settings
    - [x] /indexes/../updates
    - [x] /indexes/../stats
    - [x] /indexes/../
    - [x] /indexes
    - [x] /keys/..
    - [x] /keys
on actions:
    - [x] `search`
    - [x] `documents.add`
    - [x] `documents.get`
    - [x] `documents.delete`
    - [x] `indexes.add`
    - [x] `indexes.get`
    - [x] `indexes.update`
    - [x] `indexes.delete`
    - [x] `tasks.get`
    - [x] `settings.get`
    - [x] `settings.update`
    - [x] `settings.reset`
    - [x] `stats`
    - [x] ` dumps`

### Dump & snapshots

- [ ] ensure that keys are dumped and loaded correctly
- [ ] ensure that keys are snapshoted and loaded correctly

### PR DESCRIPTION
**Pull request by ManyTheFish**
# Pull Request

## API keys management and authorizations
Fixes #1867

## PR checklist

- [x] Test `/keys` routes
- [x] Implement `/keys` routes
- [x] Test API key authorization
- [x] Implement API key authorization
- [x] Rename Authentication header
- [x] default key creation

## Postponed in another PR
- dumps  (waiting for #1796)
- snapshot  (waiting for #1796)

### CODE REVIEW

File: meilisearch-lib/src/auth_controller/auth_store.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    type DItem = T;
+
+    fn bytes_decode(bytes: &'a [u8]) -> Option<Self::DItem> {
+        serde_json::from_slice(bytes).ok()
+    }
+}
+
+impl<'a, T> heed::BytesEncode<'a> for SerdeJsonCodec<T>
+where
+    T: Serialize + 'a,
```
Comment (MarinPostma) [inline meilisearch-lib/src/auth_controller/auth_store.rs:196]: why not using https://docs.rs/heed/0.1.0/heed/types/struct.SerdeJson.html ?

File: meilisearch-lib/src/auth_controller/auth_store.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+    fn bytes_encode(item: &Self::EItem) -> Option<Cow<[u8]>> {
+        serde_json::to_vec(item).map(|bytes| Cow::Owned(bytes)).ok()
+    }
+}
+
+pub struct ActionKeyIdCodec;
+
+impl<'a> heed::BytesDecode<'a> for ActionKeyIdCodec {
+    type DItem = ([u8; KEY_ID_LENGTH], Action, Option<&'a [u8]>);
```
Comment (MarinPostma) [inline meilisearch-lib/src/auth_controller/auth_store.rs:208]: can you documents the encoded/decoded items?

File: meilisearch-lib/src/auth_controller/auth_store.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    if mid <= slice.len() {
+        Some(slice.split_at(mid))
+    } else {
+        None
+    }
+}
+
+/// Divides one slice into an array and the tail at an index,
+/// returns `None` if `N` is out of bounds.
+pub fn try_split_array_at<T, const N: usize>(slice: &[T]) -> Option<([T; N], &[T])>
```
Comment (MarinPostma) [inline meilisearch-lib/src/auth_controller/auth_store.rs:250]: any reason why you need to own that slice here?

Comment (MarinPostma) [reply meilisearch-lib/src/auth_controller/auth_store.rs:250]: clever trick btw

File: meilisearch-lib/src/auth_controller/mod.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    let mut rng = rand::thread_rng();
+    let mut bytes = [0; KEY_ID_LENGTH];
+    for byte in bytes.iter_mut() {
+        *byte = CHARSET[rng.gen_range(0..CHARSET.len())];
+    }
+
+    bytes
+}
+
+pub fn generate_key(master_key: &[u8], uid: &[u8]) -> String {
```
Comment (MarinPostma) [inline meilisearch-lib/src/auth_controller/mod.rs:341]: can we pass uid as a &str here, so we don't have to trust the caller to pass us a valid utf8 string?

File: meilisearch-http/src/routes/mod.rs [hunk 5]
```diff
@@ -266,108 +270,6 @@ struct KeysResponse {
...
-        private: api_keys.private,
-        public: api_keys.public,
-    })
-}
-
 pub async fn get_health() -> Result<HttpResponse, ResponseError> {
     Ok(HttpResponse::Ok().json(serde_json::json!({ "status": "available" })))
 }
-
-#[cfg(test)]
```
Comment (MarinPostma) [inline meilisearch-http/src/routes/mod.rs:281]: I am quite sad to see that go. It statically checked that the routes were defined with the correct authentication, and this doesn't seem to be checked anywhere else...

**Review by MarinPostma — CHANGES_REQUESTED** [review]
I think we should put the authentication module in it's own crate rather than in meilisearch-lib, since it doesn't share a lot with it, wdyt?

File: meilisearch-lib/src/auth_controller/auth_store.rs (hunk 3 above)
Comment (ManyTheFish) [reply meilisearch-lib/src/auth_controller/auth_store.rs:250]: Just making sure that's a constant size, I don't really need to own it.

File: meilisearch-http/src/routes/mod.rs (hunk 5 above)
Comment (ManyTheFish) [reply meilisearch-http/src/routes/mod.rs:281]: I can make mistakes but I think the [authorization test file](https://github.com/meilisearch/MeiliSearch/pull/1890/files#diff-55ad1f1508692d9ba5458690ea22eea9c10c41f5cea8aa360341e8f3a7891a31) cover it. What do you think? 🤔

File: meilisearch-lib/src/auth_controller/auth_store.rs (hunk 1 above)
Comment (ManyTheFish) [reply meilisearch-lib/src/auth_controller/auth_store.rs:196]: Yes, my bad!

File: meilisearch-lib/src/auth_controller/auth_store.rs (hunk 3 above)
Comment (MarinPostma) [reply meilisearch-lib/src/auth_controller/auth_store.rs:222]: wouldn't that work?
```rust
pub fn try_split_array_at<T, const N: usize>(slice: &[T]) -> Option<(&[T; N], &[T])> {
    let (head, tail) = try_split_at(slice, N)?;
    let head = head.try_into().ok()?;
    Some((head, tail))
}
```

**Comment by ManyTheFish** [comment]
> I think we should put the authentication module in it's own crate rather than in meilisearch-lib, since it doesn't share a lot with it, wdyt?

@MarinPostma I'm a bit worried about the integration of auth in dumps and snapshots, but I agree with you.

File: meilisearch-http/src/extractors/authentication/mod.rs [hunk 6]
```diff
@@ -152,32 +48,99 @@ impl<P: Policy + 'static, D: 'static + Clone> FromRequest for GuardedData<P, D>
...
+                        Some(mut type_token) => match type_token.next() {
+                            Some("Bearer") => {
+                                // TODO: find a less hardcoded way?
+                                let index = req.match_info().get("index_uid");
+                                let token = type_token.next().unwrap_or("unknown");
+                                if P::authenticate(
+                                    auth.clone(),
+                                    token.as_bytes(),
+                                    index.map(|i| i.as_bytes()),
+                                ) {
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch-http/src/extractors/authentication/mod.rs:70]: I don't understand why we should do an `AuthContext` with a `from_request` method. 🤔

[the remark this answers is not in the document]
Comment (MarinPostma) [reply meilisearch-http/src/extractors/authentication/mod.rs:70]: you pass the request to the AuthContext so it builds itself outside of the extractor context

**Comment by MarinPostma** [comment]
bors try

**Review by MarinPostma — APPROVED** [review]
Bors merge

**Comment by MarinPostma** [comment]
bors merge

**Pull request merged**